### PR TITLE
fix(image-source): dynamically update ImageSource coordinates

### LINKED ISSUE #1013: ImageSource coordinates do not update
**Describe the bug**
I would like to update the coordinates of the ImageSource when the image URL changes.
The image does change when the URL changes, but the coordinates do not change. I have modified the heartbeat function in the example ImageOverlay component and also added the constant variable coordQuads which contains two different sets of coordinates.

**To Reproduce**
Clone maps example app.
Copy new ImageOverlay.js code (below). 
Run yarn android.

Example Code (ImageOverlay.js):
```
import React from 'react';
import MapboxGL from '@react-native-mapbox-gl/maps';

import sheet from '../styles/sheet';
import radar0 from '../assets/radar.png';
import radar1 from '../assets/radar1.png';
import radar2 from '../assets/radar2.png';

import Page from './common/Page';
import BaseExamplePropTypes from './common/BaseExamplePropTypes';

const frames = [radar0, radar1, radar2];
const coordQuads = [
  [
    [-80.425, 46.437], // top left
    [-71.516, 46.437], // top right
    [-71.516, 37.936], // bottom right
    [-80.425, 37.936], // bottom left
  ],
  [
    [-85.425, 45.437], // top left
    [-75.516, 45.437], // top right
    [-75.516, 36.936], // bottom right
    [-85.425, 36.936], // bottom left
  ],
];

class ImageOverlay extends React.Component {
  static propTypes = {
    ...BaseExamplePropTypes,
  };

  constructor(props) {
    super(props);

    this.state = {
      radarFrameIndex: 0,
      coords: coordQuads[0],
    };

    this._timeout = null;
  }

  componentDidMount() {
    this.heartbeat();
  }

  heartbeat() {
    this._timeout = setTimeout(() => {
      requestAnimationFrame(() => {
        let nextFrame = this.state.radarFrameIndex + 1;
        if (nextFrame > 1) {
          nextFrame = 0;
        }
        this.setState({radarFrameIndex: nextFrame, coords: coordQuads[nextFrame]})
        this.heartbeat();
      });
    }, 500);
  }

  componentWillUnmount() {
    if (this._timeout) {
      clearTimeout(this._timeout);
    }
  }

  render() {
    return (
      <Page {...this.props}>
        <MapboxGL.MapView
          ref={(ref) => (this.map = ref)}
          style={sheet.matchParent}
          styleURL={MapboxGL.StyleURL.Satellite}>
          <MapboxGL.Camera
            zoomLevel={5.2}
            centerCoordinate={[-75.789, 41.874]}
          />

          <MapboxGL.Animated.ImageSource
            key="d"
            id="radarSource"
            coordinates={this.state.coords}
            url={frames[this.state.radarFrameIndex]}>
            <MapboxGL.RasterLayer id="radarLayer"
            style={{rasterOpacity: .6}} />
          </MapboxGL.Animated.ImageSource>
        </MapboxGL.MapView>
      </Page>
    );
  }
}

export default ImageOverlay;

```
**Expected behavior**
I expect the ImageSource to change shapes as the heartbeat function continues to run.
Only the image changes (as expected since the URL is changing) and the coordinates stay the same.
I have logged the coordinates (this.state.coords) and they are being updated.

**Versions (please complete the following information):**
 - Platform: Android
 - Device: Pixel 2
 - Emulator/ Simulator: yes
 - OS: Android 28
 - react-native-mapbox-gl Version: 6.4.1
 - React Native Version: 0.62.1

**Additional context**
I have also run this on a physical device.
Samsung Galaxy Tab A

Thanks for your time.

### PR DESCRIPTION
**Pull request by ferdicus**
This fixes #1013 

Additionally extends the example with a button that switches between dynamic and static coordinates.

imageSource_dynamic_coord

### CODE REVIEW

**Comment by ferdicus** [comment]
@mfazekas, if you find the time, can you have a look at the iOS side of things here please? 

Thanks 🙇

**Comment by ferdicus** [comment]
@mfazekas, I've added some hacked together code, which gets the job done.
Unfortunately this only seems to work now with the dynamically updating coordinates/ urls, but not with the actual ImageOverlay example 😭 


In order to test this, simple replace the content of `example ImageOverlay.js`
With the code here:


<details> 

```jsx

import React from 'react';
import MapboxGL from '@react-native-mapbox-gl/maps';

import sheet from '../styles/sheet';
import radar0 from '../assets/radar.png';
import radar1 from '../assets/radar1.png';
import radar2 from '../assets/radar2.png';

import Page from './common/Page';
import BaseExamplePropTypes from './common/BaseExamplePropTypes';

const frames = [radar0, radar1, radar2];
const coordQuads = [
  [
    [-80.425, 46.437], // top left
    [-71.516, 46.437], // top right
    [-71.516, 37.936], // bottom right
    [-80.425, 37.936], // bottom left
  ],
  [
    [-85.425, 45.437], // top left
    [-75.516, 45.437], // top right
    [-75.516, 36.936], // bottom right
    [-85.425, 36.936], // bottom left
  ],
];

class ImageOverlay extends React.Component {
  static propTypes = {
    ...BaseExamplePropTypes,
  };

  constructor(props) {
    super(props);

    this.state = {
      radarFrameIndex: 0,
      coords: coordQuads[0],
    };

    this._timeout = null;
  }

  componentDidMount() {
    this.heartbeat();
  }

  heartbeat() {
    this._timeout = setTimeout(() => {
      requestAnimationFrame(() => {
        let nextFrame = this.state.radarFrameIndex + 1;
        if (nextFrame > 1) {
          nextFrame = 0;
        }
        this.setState({radarFrameIndex: nextFrame, coords: coordQuads[nextFrame]})
        this.heartbeat();
      });
    }, 500);
  }

  componentWillUnmount() {
    if (this._timeout) {
      clearTimeout(this._timeout);
    }
  }

  render() {
    return (
      <Page {...this.props}>
        <MapboxGL.MapView
          ref={(ref) => (this.map = ref)}
          style={sheet.matchParent}
          styleURL={MapboxGL.StyleURL.Satellite}>
          <MapboxGL.Camera
            zoomLevel={4}
            centerCoordinate={[-75.789, 41.874]}
          />

          <MapboxGL.Animated.ImageSource
            key="d"
            id="radarSource"
            coordinates={this.state.coords}
            url={frames[this.state.radarFrameIndex]}>
            <MapboxGL.RasterLayer id="radarLayer"
            style={{rasterOpacity: .6}} />
          </MapboxGL.Animated.ImageSource>
        </MapboxGL.MapView>
      </Page>
    );
  }
}

export default ImageOverlay;

```
</details>

Would appreciate some pointers, thanks in advance 🙇

File: ios/RCTMGL/RCTMGLImageSource.m [hunk 1]
```diff
@@ -21,6 +21,21 @@ - (void)setUrl:(NSString *)url
...
+    CLLocationCoordinate2D bottomLeft = CLLocationCoordinate2DMake([coordinates[3][1] floatValue], [coordinates[3][0] floatValue]);
+
+    MGLCoordinateQuad _coordinates = MGLCoordinateQuadMake(topLeft, bottomLeft, bottomRight, topRight);
+    
+    if (self.source != nil) {
+        MGLImageSource *source = (MGLImageSource *)self.source;
+        source.coordinates = _coordinates;
+    }
+}
+
```
Comment (mfazekas) [inline ios/RCTMGL/RCTMGLImageSource.m:38]:
```suggestion
- (void)setCoordinates:(NSArray<NSArray<NSNumber *> *> *)coordinates
{
    _coordinates = coordinates;
    if (self.source != nil) {
        MGLImageSource *source = (MGLImageSource *)self.source;
        source.coordinates = [self _makeCoordQuad];
    }
}
```

**Review by mfazekas — CHANGES_REQUESTED** [review]

File: ios/RCTMGL/RCTMGLImageSource.m (hunk 1 above)
Comment (ferdicus) [reply ios/RCTMGL/RCTMGLImageSource.m:38]: Nice, this does it - thanks! 

I'll enhance the example and merge without another review, if you're fine with that. 

Cheers

Comment (mfazekas) [reply ios/RCTMGL/RCTMGLImageSource.m:38]: Sure fine!

Comment (mfazekas) [reply ios/RCTMGL/RCTMGLImageSource.m:38]: 👍

**Pull request merged**